### PR TITLE
log existing errors before overwrite in granule records on file write failure

### PR DESCRIPTION
**Pull request by nemreid**
**Summary:** Summary of changes

Addresses [CUMULUS-XX: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-XXX)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js [hunk 1]
```diff
@@ -215,7 +215,7 @@ describe('The Ingest Granule failure workflow', () => {
 
       expect(granule.status).toBe('failed');
       expect(granule.error.Error).toBe(syncGranFailedDetail.error);
-      expect(JSON.parse(granule.error.Cause)).toEqual(JSON.parse(syncGranFailedDetail.cause));
+      expect(granule.error.Cause).toEqual(JSON.parse(syncGranFailedDetail.cause));
```
Comment (nemreid) [inline example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js]: The alternative to this PR would be deleting these two lines.

File: example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js [hunk 2]
```diff
@@ -214,8 +214,8 @@ describe('The Ingest Granule failure workflow', () => {
       const granule = JSON.parse(granuleResponse.body);
 
       expect(granule.status).toBe('failed');
-      expect(granule.error.Error).toBe(syncGranFailedDetail.error);
-      expect(JSON.parse(granule.error.Cause)).toEqual(JSON.parse(syncGranFailedDetail.cause));
+      expect(granule.error.Error).toBeDefined();
```
Comment (Jkovarik) [inline example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js:217]: For posterity - per pairing - we're relying on units to cover the *specific* error here - that one exists implies a failure (1/2 possible) was recorded for purposes of the integration test.

Comment (Jkovarik) [inline example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js:217]: The title of this test should be updated to be more ... exact about what we're testing.

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js [hunk 3]
```diff
@@ -453,7 +454,8 @@ const _writeGranule = async ({
   await _writeGranuleFiles({
     files,
     granuleCumulusId,
-    granuleId: granule.granuleId,
+    granule,
+    workflowError: error,
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js:458]: Nit: _writeGranuleFiles's docstring is out of date.   Presumably from a prior PR in this sequence...?

Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js:458]: Also does this work right in the underlying method, given `error` is generated from `writeGranules`'s call to `parseException` which does:

```
function parseException(exception) {
  if (isNil(exception)) return {};
  if (isObject(exception)) return exception;
  return {
    Error: 'Unknown Error',
    Cause: exception,
  };
}
```

?

I'll comment on the line above as well.

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js [hunk 4]
```diff
@@ -306,18 +307,18 @@ const _writeGranuleFiles = async ({
       knex,
     });
   } catch (error) {
-    log.error('Failed writing some files to Postgres', error);
-
+    if (workflowError !== undefined) {
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js:310]: I don't believe this works - error can be `{}` as a result of writeGranules's call to parseException.       

File this under 'I really want us to get the API code into TS ASAP'

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Changes look good, one thing to fix.

File: packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 const AggregateError = require('aggregate-error');
+const isEmpty = require('lodash/isEmpty');
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js:4]: 👍🏻  isEmpty works.

**Review by Jkovarik — APPROVED** [review]
My feedback has been addressed, nice work!

**Pull request merged**